R/dictionary.R

```diff
@@ -0,0 +1,169 @@
+#' Create a dictionary.
+#'
+#' @description
+#'
+#' Dictionaries are a concept of types modelled after R
+#' environments. Dictionaries are containers of R objects that:
+#'
+#' - Contain uniquely named objects.
+#'
+#' - Can only be indexed by name. They must implement the extracting
+#'   operators `$` and `[[`. The latter returns an error when indexed
+#'   by position because dictionaries are not vectors (they are
+#'   unordered).
+#'
+#' - Report a clear error message when asked to extract a name that
+#'   does not exist. This error message can be customised with the
+#'   `lookup_msg` constructor argument.
+#'
+#' @details
+#'
+#' Dictionaries are used within the tidy evaluation framework for
+#' creating pronouns that can be explicitly referred to from captured
+#' code. See [eval_tidy()].
+#'
+#' @param x An object for which you want to find associated data.
+#' @param lookup_msg An error message when your data source is
+#'   accessed inappropriately (by position rather than name).
+#' @param read_only Whether users can replace elements of the
+#'   dictionary.
+#' @name dictionary
+#' @export
+as_dictionary <- function(x, lookup_msg = NULL, read_only = FALSE) {
+  UseMethod("as_dictionary")
+}
+#' @export
+as_dictionary.default <- function(x, lookup_msg = NULL, read_only = FALSE) {
+  x <- discard_unnamed(x)
+  if (!is_dictionarish(x)) {
+    abort("Data source must be a dictionary")
+  }
+  new_dictionary(as.list(x), lookup_msg, read_only)
+}
+#' @export
+as_dictionary.dictionary <- function(x, lookup_msg = NULL, read_only = FALSE) {
+  dict <- unclass_dict(x)
+  dict$lookup_msg <- lookup_msg %||% x$lookup_msg
+  dict$read_only <- read_only
+  set_attrs(dict, class = class(x))
+}
+#' @export
+as_dictionary.NULL <- function(x, lookup_msg = NULL, read_only = FALSE) {
+  new_dictionary(list(), lookup_msg, read_only)
+}
+#' @export
+as_dictionary.environment <- function(x, lookup_msg = NULL, read_only = FALSE) {
+  lookup_msg <- lookup_msg %||% "Object `%s` not found in environment"
+  new_dictionary(x, lookup_msg, read_only)
+}
+#' @export
+as_dictionary.data.frame <- function(x, lookup_msg = NULL, read_only = FALSE) {
+  lookup_msg <- lookup_msg %||% "Variable `%s` not found in data"
```

Comment:
This should probably be `` "Column `%s` not found in data" `` to be consistent with dplyr